registry/adapters/kubernetes/kubernetes.go

```diff
@@ -16,73 +16,179 @@ package kubernetes
 
 import (
 	"fmt"
+	"reflect"
 	"strings"
 	"sync"
 	"time"
 
-	log "github.com/Sirupsen/logrus"
+	"encoding/json"
 
 	"github.com/amalgam8/amalgam8/pkg/auth"
+	"github.com/amalgam8/amalgam8/pkg/datastructures"
+	"github.com/amalgam8/amalgam8/pkg/errors"
 	"github.com/amalgam8/amalgam8/registry/api"
 	"github.com/amalgam8/amalgam8/registry/utils/logging"
+	"k8s.io/client-go/kubernetes"
+	kubeapi "k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/pkg/runtime"
+	"k8s.io/client-go/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
 )
 
 const (
-	module = "K8SADAPTER"
+	// EndpointsCacheResyncPeriod is the period in which we do a full resync of the endpoints cache.
+	EndpointsCacheResyncPeriod = time.Duration(60) * time.Second
+
+	// PodCacheResyncPeriod is the period in which we do a full resync of the endpoints cache.
+	PodCacheResyncPeriod = time.Duration(60) * time.Second
 )
 
 // Make sure we implement the ServiceDiscovery interface
 var _ api.ServiceDiscovery = (*Adapter)(nil)
 
-// Config stores configurable attributes of the kubernetes adapter
+// Package global logger
+var logger = logging.GetLogger("KUBERNETES")
+
+// Config stores configurable attributes of the Kubernetes adapter.
 type Config struct {
 	URL       string
 	Token     string
 	Namespace auth.Namespace
+
+	// Client to be used by the Kubernetes adapter.
+	// If no client is provided, then a client is created
+	// according the specified URL/Token/Namespace, if provided,
+	// or from the local service account, if running within a Kubernetes pod.
+	Client kubernetes.Interface
 }
 
-// Adapter for Eureka Service Discovery
+// Adapter for Kubernetes Service Discovery.
 type Adapter struct {
-	namespace auth.Namespace
-	client    *client
+	endpointsCache      cache.Store
+	endpointsController cache.ControllerInterface
+
+	podCache      cache.Store
+	podController cache.ControllerInterface
 
+	// services maps a service name to a list of service instances
 	services map[string][]*api.ServiceInstance
 
-	logger *log.Entry
-	sync.RWMutex
+	// servicePods maps a service name to a set of pods implementing it
+	servicePods map[string]datastructures.StringSet
+
```

Comment:
`api.ServiceInstance` is a8's representation of an instance of a service. It is built and stored in the `services` map, so that it won't have to be recalculated with every `ListInstances()` call.
`servicePods` maps service name to a set of pod UIDs that implement it. This is maintained, together with the opposite mapping (pod UID --> services), so that we can supporting dynamic labels update on pods.